src/brevitas/graph/base.py

```diff
@@ -174,6 +178,112 @@ def apply(self, graph_model: GraphModule) -> GraphModule:
         return graph_model
 
 
+class ModuleInstanceRegisterParametrization(Transform):
+
+    def __init__(
+            self, old_module_instance: Module, tensor_name: str,
+            parametrization_module: Module) -> None:
+        self.old_module_instance = old_module_instance
+        self.tensor_name = tensor_name
+        self.parametrization_module = parametrization_module
+
+    def apply(self, model: GraphModule) -> GraphModule:
+        for old_module in model.modules():
+            if old_module is self.old_module_instance:
+                # register the parametrization in the old_module
+                parametrize.register_parametrization(
+                    old_module, self.tensor_name, self.parametrization_module, unsafe=True)
+                break
+        return model
+
+
+class RotationWeightParametrization(torch.nn.Module):
+
+    def __init__(
+        self,
+        rot_mat: torch.nn.Parameter,
+        rot_func: Callable,
+        axis: int,
+        K: Optional[int] = None,
+    ) -> None:
+        super().__init__()
+        self.rot_mat = rot_mat
+        self.rot_func = rot_func
+        self.axis = axis
+        self.K = K
+
+    def forward(self, weight: torch.Tensor) -> torch.Tensor:
+
+        if self.axis == 0:
+            weight = self.rot_func(weight.t(), self.rot_mat, self.K).t()
+        elif self.axis == 1:
+            weight = self.rot_func(weight, self.rot_mat, self.K)
+        else:
+            raise RuntimeError("Not supported yet")
+
+        return weight
+
+
+class ModuleInstanceFuseRotationWeights(Transform):
+
+    def __init__(
+        self,
+        old_module_instance: Module,
+        rot_mat: Union[Parameter, Tensor],
+        rot_func: Callable,
+        K: Optional[int],
+        tensor_name: str,
+        axis: int,
+    ):
+        self.old_module_instance = old_module_instance
+        self.rot_mat = rot_mat
+        self.rot_func = rot_func
+        self.K = K
+        self.tensor_name = tensor_name
+        self.axis = axis
+
+    def apply(self, model: GraphModule) -> GraphModule:
+        for old_module in model.modules():
+            if old_module is self.old_module_instance:
+                if hasattr(old_module, 'allocate_params'):
```

Comment:
Add a comment, maybe TODO, whether this should live here or outside the apply function. I'm not sure about it